<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Introduction" data-toc-modified-id="Introduction-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Introduction</a></span></li><li><span><a href="#Features" data-toc-modified-id="Features-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Features</a></span></li><li><span><a href="#Methodology" data-toc-modified-id="Methodology-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Methodology</a></span></li><li><span><a href="#Code" data-toc-modified-id="Code-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Code</a></span><ul class="toc-item"><li><span><a href="#Import-Libraries" data-toc-modified-id="Import-Libraries-4.1"><span class="toc-item-num">4.1&nbsp;&nbsp;</span>Import Libraries</a></span></li><li><span><a href="#Data-ingestion-and-storage" data-toc-modified-id="Data-ingestion-and-storage-4.2"><span class="toc-item-num">4.2&nbsp;&nbsp;</span>Data ingestion and storage</a></span></li><li><span><a href="#Data-summarization" data-toc-modified-id="Data-summarization-4.3"><span class="toc-item-num">4.3&nbsp;&nbsp;</span>Data summarization</a></span></li><li><span><a href="#Data-cleaning-and-preprocessing" data-toc-modified-id="Data-cleaning-and-preprocessing-4.4"><span class="toc-item-num">4.4&nbsp;&nbsp;</span>Data cleaning and preprocessing</a></span></li><li><span><a href="#Modeling" data-toc-modified-id="Modeling-4.5"><span class="toc-item-num">4.5&nbsp;&nbsp;</span>Modeling</a></span></li></ul></li><li><span><a href="#Conclusion" data-toc-modified-id="Conclusion-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Conclusion</a></span></li><li><span><a href="#References:" data-toc-modified-id="References:-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>References:</a></span></li></ul></div>

# Kaggle Cases : Project - 3 : Multi-class classification

Get out your dowsing rods, electromagnetic sensors, … and gradient boosting machines. Kaggle is haunted and we need your help. After a month of making scientific observations and taking careful measurements, we’ve determined that 900 ghouls, ghosts, and goblins are infesting our halls and frightening our data scientists. When trying garlic, asking politely, and using reverse psychology didn't work, it became clear that machine learning is the only answer to banishing our unwanted guests.

The link to the dataset can be found in the references[1]


## Introduction

So now the hour has come to put the data we’ve collected in your hands. We’ve managed to identify 371 of the ghastly creatures, but need your help to vanquish the rest. And only an accurate classification algorithm can thwart them. Use bone length measurements, severity of rot, extent of soullessness, and other characteristics to distinguish (and extinguish) the intruders. Are you ghost-busters up for the challenge?

We are to vanquish the ghost haunting Kagglers :) 

GHOST-BUSTERS :P


## Features

id - id of the creature

bone_length - average length of bone in the creature, normalized between 0 and 1

rotting_flesh - percentage of rotting flesh in the creature

hair_length - average hair length, normalized between 0 and 1

has_soul - percentage of soul in the creature

color - dominant color of the creature: 'white','black','clear','blue','green','blood'

type - target variable: 'Ghost', 'Goblin', and 'Ghoul'


## Methodology 


1. Data ingestion and storage 

2. Data summarization

3. Data cleaning and prcoessing

4. Data modeling


## Code

### Import Libraries

In [31]:
import pandas as pd
import numpy as np

from sklearn.model_selection import train_test_split
from sklearn.model_selection import GridSearchCV
from sklearn.ensemble import RandomForestClassifier
import matplotlib.pyplot as plt
from sklearn.preprocessing import StandardScaler

%matplotlib inline


### Data ingestion and storage

Data is downloaded from the link shared on the canvas for the Project.
Then its loaded as a Dataframe object, using the Pandas library in Python.


In [9]:
df_train = pd.read_csv('./data/train.csv')
print('Dimension : ', df_train.shape)
df_train.head()

Dimension :  (371, 7)


id  bone_length  rotting_flesh  hair_length  has_soul  color    type
0   0     0.354512       0.350839     0.465761  0.781142  clear   Ghoul
1   1     0.575560       0.425868     0.531401  0.439899  green  Goblin
2   2     0.467875       0.354330     0.811616  0.791225  black   Ghoul
3   4     0.776652       0.508723     0.636766  0.884464  black   Ghoul
4   5     0.566117       0.875862     0.418594  0.636438  green   Ghost

In [10]:
df_test = pd.read_csv('./data/test.csv')
print('Dimension : ', df_test.shape)
df_test.head()


Dimension :  (529, 6)


id  bone_length  rotting_flesh  hair_length  has_soul  color
0   3     0.471774       0.387937     0.706087  0.698537  black
1   6     0.427332       0.645024     0.565558  0.451462  white
2   9     0.549602       0.491931     0.660387  0.449809  black
3  10     0.638095       0.682867     0.471409  0.356924  white
4  13     0.361762       0.583997     0.377256  0.276364  black

### Data summarization

Below we look at the data, each column names the count of the rows. We also describe partcular columns that we are interested in using the describe function and finally select multiple columns from all the columns.

In [12]:
df_train.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 371 entries, 0 to 370
Data columns (total 7 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   id             371 non-null    int64  
 1   bone_length    371 non-null    float64
 2   rotting_flesh  371 non-null    float64
 3   hair_length    371 non-null    float64
 4   has_soul       371 non-null    float64
 5   color          371 non-null    object 
 6   type           371 non-null    object 
dtypes: float64(4), int64(1), object(2)
memory usage: 20.4+ KB


In [13]:
df_train.describe()

id  bone_length  rotting_flesh  hair_length    has_soul
count  371.000000   371.000000     371.000000   371.000000  371.000000
mean   443.676550     0.434160       0.506848     0.529114    0.471392
std    263.222489     0.132833       0.146358     0.169902    0.176129
min      0.000000     0.061032       0.095687     0.134600    0.009402
25%    205.500000     0.340006       0.414812     0.407428    0.348002
50%    458.000000     0.434891       0.501552     0.538642    0.466372
75%    678.500000     0.517223       0.603977     0.647244    0.600610
max    897.000000     0.817001       0.932466     1.000000    0.935721

In [14]:
df_train.isnull().sum() ## No missing values

id               0
bone_length      0
rotting_flesh    0
hair_length      0
has_soul         0
color            0
type             0
dtype: int64

In [15]:
df_test.isnull().sum() ## No missing values

id               0
bone_length      0
rotting_flesh    0
hair_length      0
has_soul         0
color            0
dtype: int64

### Data cleaning and preprocessing

Here we do a bit of data cleaning.

We create dumy variable for Color


In [22]:
# converting labels to number
df_train.type[df_train.type == 'Ghoul'] = 1
df_train.type[df_train.type == 'Goblin'] = 2
df_train.type[df_train.type == 'Ghost'] = 3

C:\Users\Prashita\anacondaMay\lib\site-packages\ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\Users\Prashita\anacondaMay\lib\site-packages\ipykernel_launcher.py:3: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  This is separate from the ipykernel package so we can avoid doing imports until
C:\Users\Prashita\anacondaMay\lib\site-packages\ipykernel_launcher.py:4: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  a

In [23]:
df_train.head()

id  bone_length  rotting_flesh  hair_length  has_soul  color type
0   0     0.354512       0.350839     0.465761  0.781142  clear    1
1   1     0.575560       0.425868     0.531401  0.439899  green    2
2   2     0.467875       0.354330     0.811616  0.791225  black    1
3   4     0.776652       0.508723     0.636766  0.884464  black    1
4   5     0.566117       0.875862     0.418594  0.636438  green    3

In [24]:
X = pd.get_dummies(df_train, columns=['color'], prefix=['color'], prefix_sep='_')
X.head()

id  bone_length  rotting_flesh  hair_length  has_soul type  color_black  \
0   0     0.354512       0.350839     0.465761  0.781142    1            0   
1   1     0.575560       0.425868     0.531401  0.439899    2            0   
2   2     0.467875       0.354330     0.811616  0.791225    1            1   
3   4     0.776652       0.508723     0.636766  0.884464    1            1   
4   5     0.566117       0.875862     0.418594  0.636438    3            0   

   color_blood  color_blue  color_clear  color_green  color_white  
0            0           0            1            0            0  
1            0           0            0            1            0  
2            0           0            0            0            0  
3            0           0            0            0            0  
4            0           0            0            1            0

In [50]:
# processing test data
test_ids = df_test.id

final_test = pd.get_dummies(df_test, columns=['color'], prefix=['color'], prefix_sep='_')
final_test = final_test.drop(columns=['id'])

final_test.head(5)


bone_length  rotting_flesh  hair_length  has_soul  color_black  \
0     0.471774       0.387937     0.706087  0.698537            1   
1     0.427332       0.645024     0.565558  0.451462            0   
2     0.549602       0.491931     0.660387  0.449809            1   
3     0.638095       0.682867     0.471409  0.356924            0   
4     0.361762       0.583997     0.377256  0.276364            1   

   color_blood  color_blue  color_clear  color_green  color_white  
0            0           0            0            0            0  
1            0           0            0            0            1  
2            0           0            0            0            0  
3            0           0            0            0            1  
4            0           0            0            0            0

### Modeling

In [42]:
Y = X.type.astype(int)
Xnew = X.drop(['type','id'], axis = 1)
Xnew.head()

bone_length  rotting_flesh  hair_length  has_soul  color_black  \
0     0.354512       0.350839     0.465761  0.781142            0   
1     0.575560       0.425868     0.531401  0.439899            0   
2     0.467875       0.354330     0.811616  0.791225            1   
3     0.776652       0.508723     0.636766  0.884464            1   
4     0.566117       0.875862     0.418594  0.636438            0   

   color_blood  color_blue  color_clear  color_green  color_white  
0            0           0            1            0            0  
1            0           0            0            1            0  
2            0           0            0            0            0  
3            0           0            0            0            0  
4            0           0            0            1            0

In [43]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(Xnew, Y, test_size=0.2, random_state=42)

In [44]:
print(X_train.shape, y_train.shape)
print(X_test.shape, y_test.shape)

(296, 10) (296,)
(75, 10) (75,)


In [45]:
# scaling data
scaler = StandardScaler()
X_train = scaler.fit_transform(X_train)
X_test = scaler.transform(X_test)

In [46]:
from sklearn.ensemble import RandomForestClassifier
clf = RandomForestClassifier()
param_grid={
    'n_estimators':[10,30,50,100],
    'max_depth':[3,6,10,None]
}
gridsearch_rf = GridSearchCV(clf,param_grid=param_grid,scoring='accuracy',n_jobs=4,cv=4,verbose=3).fit(X_train,y_train)
best_rf = gridsearch_rf.best_estimator_

Fitting 4 folds for each of 16 candidates, totalling 64 fits


[Parallel(n_jobs=4)]: Using backend LokyBackend with 4 concurrent workers.
[Parallel(n_jobs=4)]: Done  40 tasks      | elapsed:    1.6s
[Parallel(n_jobs=4)]: Done  64 out of  64 | elapsed:    3.1s finished


In [47]:
# train accuracy
best_rf.score(X_train, y_train)


0.8344594594594594

In [48]:
# test accuracy
best_rf.score(X_test, y_test)

0.68

In [51]:
# final predictions
predictions = best_rf.predict(final_test)

In [53]:
result = pd.DataFrame({
    "id":test_ids,
    "type":predictions
})

# converting back to labesl
result.type[result.type == 1 ] = 'Ghoul'
result.type[result.type == 2 ] = 'Goblin'
result.type[result.type == 3 ] = 'Ghost'

print(result.shape)
result.head(5)

(529, 2)


C:\Users\Prashita\anacondaMay\lib\site-packages\ipykernel_launcher.py:7: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  import sys


id    type
0   3   Ghoul
1   6   Ghoul
2   9   Ghoul
3  10   Ghoul
4  13  Goblin

## Conclusion

Thus the whole project was a learning experience, as it provided a platform to work with an interesting dataset and carry out multi class classification.

We used the Random Forest classifier to detect whether it is a Ghoul, Ghost or Goblin.

We ontained a test accuracy of 68% and train accuracy of 83.45%.

 

## References:

1. https://www.kaggle.com/c/ghouls-goblins-and-ghosts-boo/data
